keps/sig-storage/1845-prioritization-on-volume-capacity/README.md

```diff
@@ -0,0 +1,505 @@
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With [volume topology-aware
+scheduling](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md),
+the scheduler will filter nodes considering volume topology constraints.
+Unbound PVCs will be bound with the smallest fit PVs in a topology, however, the
+scheduler does not try to prioritize nodes in different topologies based on
+volume capacity. This leads to a waste of resources which we should avoid.
+
+This KEP continues the work proposed [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md#priority).
+
+## Motivation
+
+Because we didn't take volume capacity into account when scheduling a pod that
+can run in multiple topologies. A large PV may be used by a PVC with a small
+capacity request, even if there are many suitable small PVs in other
+topologies. PVCs with a large capacity request may not find feasible PVs to use
+if too many large PVs are consumed by PVCs with a small capacity request.
+
+Like CPU, Memory resources, the scheduler should take volume capacity into
+account in scheduling pods to ensure balanced resource usage.
+
+### Goals
+
+- Nodes of topology in which volumes with close capacity are preferred in
+  scheduling
+- Support arbitrary PV topology domain (e.g. hostname, zone, rack)
+
+### Non-Goals
+
+- Prioritizing nodes based on the total available capacity of static provisioned
+  PVs
+- Prioritizing nodes based on the total available number of static provisioned
+  PVs
+- Prioritizing nodes based on the total available capacity for dynamic
+  provisioning PVCs
```

Comment:
To clarify: dynamic provisioning always provision the exact storage size from the PVC request, so we don't consider it, right?

Reply:
yes, although the provisioned size could be larger than the request size due to implementation limitations (e.g. minimum disk/unit size)

Reply:
the constraints are on PVs (from which node(s) the PV can be accessed, specified in `PV.Spec.NodeAffinity`). 

in dynamic provisioning, there are no constraints except the allowed topologies specified in the storage class. if all zones are allowed, any zone can be chosen, but once the PV is provisioned in the PVC, the pod using this PVC can not be rescheduled to another zone.

 > can that be specified in the claim?

currently, no. in the future, once the [storage capacity tracking](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1472-storage-capacity-tracking) is ready, the total available capacity in each topology can be taken into account and the request capacity of the claim will be used